src/chainstate/stacks/index/proofs.rs

```diff
@@ -228,7 +234,129 @@ impl Deref for TrieMerkleProof {
+            TrieMerkleProofType::Node4(_) => TrieMerkleProofTypeIndicator::Node4,
+            TrieMerkleProofType::Node16(_) => TrieMerkleProofTypeIndicator::Node16,
+            TrieMerkleProofType::Node48(_) => TrieMerkleProofTypeIndicator::Node48,
+            TrieMerkleProofType::Node256(_) => TrieMerkleProofTypeIndicator::Node256,
+            TrieMerkleProofType::Leaf(_) => TrieMerkleProofTypeIndicator::Leaf,
+            TrieMerkleProofType::Shunt(_) => TrieMerkleProofTypeIndicator::Shunt
+        } as u8;
+
+        type_byte.consensus_serialize(fd)?;
+
+        match self {
+            TrieMerkleProofType::Node4((id, proof_node, hashes)) => serialize_id_hash_node(fd, id, proof_node, hashes),
+            TrieMerkleProofType::Node16((id, proof_node, hashes)) => serialize_id_hash_node(fd, id, proof_node, hashes),
+            TrieMerkleProofType::Node48((id, proof_node, hashes)) => serialize_id_hash_node(fd, id, proof_node, hashes),
+            TrieMerkleProofType::Node256((id, proof_node, hashes)) => serialize_id_hash_node(fd, id, proof_node, hashes),
+            TrieMerkleProofType::Leaf((id, leaf_node)) => {
+                id.consensus_serialize(fd)?;
+                leaf_node.consensus_serialize(fd)
+            },
+            TrieMerkleProofType::Shunt((id, hashes)) => {
+                id.consensus_serialize(fd)?;
+                hashes.consensus_serialize(fd)
+            },
+        }
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<TrieMerkleProofType, ::net::Error> {
+        let type_byte = TrieMerkleProofTypeIndicator::from_u8(read_next(fd)?)
+            .ok_or_else(|| ::net::Error::DeserializeError("Bad type byte in Trie Merkle Proof".into()))?;
+
+        let codec = match type_byte {
+            TrieMerkleProofTypeIndicator::Node4 => {
+                TrieMerkleProofType::Node4(deserialize_id_hash_node!(fd, [TrieHash([0; 32]); 3]))
+            },
+            TrieMerkleProofTypeIndicator::Node16 => {
+                TrieMerkleProofType::Node16(deserialize_id_hash_node!(fd, [TrieHash([0; 32]); 15]))
+            },
+            TrieMerkleProofTypeIndicator::Node48 => {
+                TrieMerkleProofType::Node48(deserialize_id_hash_node!(fd, [TrieHash([0; 32]); 47]))
+            },
+            TrieMerkleProofTypeIndicator::Node256 => {
+                TrieMerkleProofType::Node256(deserialize_id_hash_node!(fd, [TrieHash([0; 32]); 255]))
+            },
+            TrieMerkleProofTypeIndicator::Leaf => {
+                let id = read_next(fd)?;
+                let leaf_node = read_next(fd)?;
+                TrieMerkleProofType::Leaf((id, leaf_node))
+            },
+            TrieMerkleProofTypeIndicator::Shunt => {
+                let id = read_next(fd)?;
+                let hashes = read_next(fd)?;
+                TrieMerkleProofType::Shunt((id, hashes))
+            },
+        };
+
+        Ok(codec)
+    }
+}
+
 impl TrieMerkleProof {
+    pub fn to_hex(&self) -> String {
```

Comment:
Because MARF proofs can get big -- i.e. nodes can be half-full, meaning that half of the child hashes will be identical (empty) -- should we consider adding a more efficient representation here?  Maybe we can compress the proof and then hex-encode it?  Or better, base64-encode it?  Doesn't have to be in this PR, but we should probably decide sooner rather than later so clients can expect it.

Reply:
I think we should use the same encoding across the board (apart from addresses). The same reason could be applied to use base64 elsewhere as well. However, HTTP compression would remove a lot of this concern.

Reply:
While I agree with that, I do think that we'll want to have a compressed representation of a MARF proof at some point before mainnet (not in this PR) if we're going to expect people to use MARF proofs in Clarity.

Reply:
I opened an issue for this: https://github.com/blockstack/stacks-blockchain/issues/1397